crates/circuit/src/packed_instruction.rs

```diff
@@ -0,0 +1,497 @@
+///
+/// # Safety
+///
+/// `PackedOperation` asserts ownership over its contained pointer (if not a `StandardGate`).  This
+/// has the following requirements:
+///
+/// * The pointer must be managed by a `Box` using the global allocator.
+/// * The pointed-to data must match the type of the discriminant used to store it.
+/// * `PackedOperation` must take care to forward implementations of `Clone` and `Drop` to the
+///   contained pointer.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct PackedOperation(usize);
+
+impl PackedOperation {
+    /// The bits representing the `PackedOperationType` discriminant.  This can be used to mask out
+    /// the discriminant, and defines the rest of the bit shifting.
+    const DISCRIMINANT_MASK: usize = 0b11;
+    /// The number of bits used to store the discriminant metadata.
+    const DISCRIMINANT_BITS: u32 = Self::DISCRIMINANT_MASK.count_ones();
+    /// A bitmask that masks out only the standard gate information.  This should always have the
+    /// same effect as `POINTER_MASK` because the high bits should be 0 for a `StandardGate`, but
+    /// this is defensive against us adding further metadata on `StandardGate` later.  After
+    /// masking, the resulting integer still needs shifting downwards to retrieve the standard gate.
+    const STANDARD_GATE_MASK: usize = (u8::MAX as usize) << Self::DISCRIMINANT_BITS;
+    /// A bitmask that retrieves the stored pointer directly.  The discriminant is stored in the
+    /// low pointer bits that are guaranteed to be 0 by alignment, so no shifting is required.
+    const POINTER_MASK: usize = usize::MAX ^ Self::DISCRIMINANT_MASK;
+
+    /// Extract the discriminant of the operation.
+    #[inline]
+    fn discriminant(&self) -> PackedOperationType {
+        ::bytemuck::checked::cast((self.0 & Self::DISCRIMINANT_MASK) as u8)
+    }
+
+    /// Get the contained pointer to the `PyGate`/`PyInstruction`/`PyOperation` that this object
+    /// contains.
+    ///
+    /// **Panics** if the object represents a standard gate; see `try_pointer`.
+    #[inline]
+    fn pointer(&self) -> NonNull<()> {
+        self.try_pointer().unwrap()
+    }
+
+    /// Get the contained pointer to the `PyGate`/`PyInstruction`/`PyOperation` that this object
+    /// contains.
+    ///
+    /// Returns `None` if the object represents a standard gate.
+    #[inline]
+    pub fn try_pointer(&self) -> Option<NonNull<()>> {
+        match self.discriminant() {
+            PackedOperationType::StandardGate => None,
+            PackedOperationType::Gate
+            | PackedOperationType::Instruction
+            | PackedOperationType::Operation => {
+                let ptr = (self.0 & Self::POINTER_MASK) as *mut ();
+                // SAFETY: `PackedOperation` can only be constructed from a pointer via `Box`, which
+                // is always non-null (except in the case that we're partway through a `Drop`).
+                Some(unsafe { NonNull::new_unchecked(ptr) })
```

Comment:
Would something like this be more useful?
```suggestion
                Some(NonNull::new(ptr).expect("Qiskit bug: Expected a non-null pointer"))
```
In both cases, the function should always exit normally. But in the second case, if the impossible does happen, the user sees it's a bug and not their fault.

Reply:
I'm not a big fan of the additional runtime check for internal Qiskit bugs here, especially since particular one wouldn't help narrow down any source of errors - that would have come before by something writing an invalid pointer into the value - and wouldn't catch the category of bugs that are "invalid pointer but _not_ null", which seems equally likely if something unknown has written into the field.

Reply:
I added an expect message in f991157574cfd9c462418366486b307e712f9059.